<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Tutorial</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Tutorial</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co">#load Lund Taxonomy Classifier</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(LundTax2023Classifier)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(grid)</span></code></pre></div>
<div id="lund-taxonomy-classifier" class="section level1">
<h1>Lund Taxonomy Classifier</h1>
<p>This package depends on a set of internal and exported functions. The
main funcitons inteded for the user are;
<code>lundtax_predict_sub</code>, <code>lundtax_calc_sigscore</code>,
<code>plot_hm_signatures</code>, <code>plot_hm_sigscores</code>,
<code>plot_ranked_score</code>. In this tutorial, we will demonstrate
how to use these functions, using the bundled data to get the most out
from this package.</p>
<div id="bundled-data" class="section level2">
<h2>Bundled Data</h2>
<p>Before we get started, let’s start with exploring some of the
relevant bunlded data objects.</p>
<div id="expression-data" class="section level3">
<h3>Expression Data</h3>
<p>This package includes an example expression data set. This data is a
subset of samples from the Sjodahl 2017 paper. Let’s first see the
dimensions of the data object.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">dim</span>(sjodahl_2017)</span></code></pre></div>
<pre><code>## [1] 15697   267</code></pre>
<p>The expression data set has 15607 rows (genes) and 267 columns
(samples). Let’s have a look at the a the first 5 rows and 5 columns to
get an idea of what the data looks like.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(sjodahl_2017[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], </span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 1. Bundled expression data subset&quot;</span>)</span></code></pre></div>
<table>
<caption>Table 1. Bundled expression data subset</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">1.CEL</th>
<th align="right">2.CEL</th>
<th align="right">3.CEL</th>
<th align="right">4.CEL</th>
<th align="right">5.CEL</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">A1CF</td>
<td align="right">4.248274</td>
<td align="right">4.153923</td>
<td align="right">4.186689</td>
<td align="right">4.877813</td>
<td align="right">4.106269</td>
</tr>
<tr class="even">
<td align="left">A2M</td>
<td align="right">8.348800</td>
<td align="right">7.969504</td>
<td align="right">10.755738</td>
<td align="right">8.467495</td>
<td align="right">8.979223</td>
</tr>
<tr class="odd">
<td align="left">A2ML1</td>
<td align="right">6.694896</td>
<td align="right">4.442973</td>
<td align="right">8.871067</td>
<td align="right">4.392010</td>
<td align="right">7.685045</td>
</tr>
<tr class="even">
<td align="left">A4GALT</td>
<td align="right">7.255671</td>
<td align="right">6.337239</td>
<td align="right">6.925670</td>
<td align="right">6.239185</td>
<td align="right">6.032115</td>
</tr>
<tr class="odd">
<td align="left">A4GNT</td>
<td align="right">3.740418</td>
<td align="right">3.830598</td>
<td align="right">3.999331</td>
<td align="right">3.813662</td>
<td align="right">3.718752</td>
</tr>
</tbody>
</table>
</div>
<div id="metadata" class="section level3">
<h3>Metadata</h3>
<p>The package also includes meta data for the expression data. The
metadata can be used to aggregate the prediction calls, allowing for
interesting association between various metadata variables and specific
subtype. Let’s explore this data set by overviewing what clinical
information is available.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">str</span>(sjodahl_2017_meta)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    267 obs. of  25 variables:
##  $ sample_id        : chr  &quot;1.CEL&quot; &quot;2.CEL&quot; &quot;3.CEL&quot; &quot;4.CEL&quot; ...
##  $ cluster_order    : int  13 162 154 11 251 55 254 278 155 101 ...
##  $ age              : num  74.7 53.6 65.5 70.8 64.2 67.2 77.3 66.2 59.6 64.1 ...
##  $ gender           : Factor w/ 2 levels &quot;Female&quot;,&quot;Male&quot;: 2 2 2 2 1 2 2 2 1 2 ...
##  $ region_cx        : Factor w/ 4 levels &quot;HBG&quot;,&quot;LUND&quot;,&quot;MALMO&quot;,..: 3 3 3 3 3 3 3 3 1 3 ...
##  $ turb_grade       : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 2 3 3 2 2 2 2 3 2 2 ...
##  $ turb_stage       : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 2 2 2 2 2 2 2 1 2 2 ...
##  $ turb_skiv        : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 1 1 1 2 2 1 2 1 2 1 ...
##  $ turb_cis         : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 1 1 1 1 1 1 1 1 1 2 ...
##  $ turb_lvi         : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 2 1 1 2 2 2 2 2 1 1 ...
##  $ turb_necros      : Factor w/ 4 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;N/A&quot;: 1 1 2 2 3 3 3 3 2 2 ...
##  $ turb_apoptos     : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 2 1 1 2 2 2 2 2 2 2 ...
##  $ variant_histology: Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 1 2 2 1 1 1 1 2 1 1 ...
##  $ cT               : Factor w/ 5 levels &quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,..: 3 2 1 2 2 3 2 2 3 2 ...
##  $ cN               : Factor w/ 3 levels &quot;0&quot;,&quot;1&quot;,&quot;N/A&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ hydroneph        : Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 2 1 1 1 1 1 2 1 1 2 ...
##  $ pTCx             : Factor w/ 10 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 6 8 5 8 4 5 1 2 8 9 ...
##  $ pN               : Factor w/ 5 levels &quot;0&quot;,&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,..: 3 1 1 1 1 1 3 1 1 3 ...
##  $ adj_chemo        : Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 2 1 1 1 1 1 1 1 1 2 ...
##  $ surv_os_event    : Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 2 2 1 2 1 2 2 2 2 2 ...
##  $ surv_os_time     : num  12.8 13.5 100.4 3 77.6 ...
##  $ surv_css_event   : Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 2 2 1 1 1 2 2 1 1 2 ...
##  $ surv_css_time    : num  12.8 13.5 100.4 3 77.6 ...
##  $ surv_pfs_event   : Factor w/ 2 levels &quot;0&quot;,&quot;1&quot;: 2 2 1 1 1 2 2 1 1 2 ...
##  $ surv_pfs_time    : num  11.6 10.6 100.4 3 77.6 ...</code></pre>
<p>If the user wants to use their own metadata, this should be in a data
frame format wiith unique sample identifiers in a column called
<code>sample_id</code>. For a more detailed explanation of the
variables, see that data set documentation.</p>
</div>
<div id="other-bundled-objects" class="section level3">
<h3>Other Bundled Objects</h3>
<p>this package also includes a variety of data objects needed for the
predictor. For example, classifiers for predicting grade and subtype,
lists with genes identifiers for making prediction calls and gene
signatures. Let’s explore a few of these.</p>
<p>The <code>gene_list</code> holds information on gene identifiers that
the predictor is based on. This object has gene identfiers in both HGNC
and Ensembl format, allowing the incoming data to be in either of these
formats.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(gene_list, <span class="dv">5</span>),</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 2. Gene List&quot;</span>)</span></code></pre></div>
<table>
<caption>Table 2. Gene List</caption>
<thead>
<tr class="header">
<th align="left">ensembl_gene_id</th>
<th align="left">hgnc_symbol</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000109576</td>
<td align="left">AADAT</td>
</tr>
<tr class="even">
<td align="left">ENSG00000183044</td>
<td align="left">ABAT</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000108846</td>
<td align="left">ABCC3</td>
</tr>
<tr class="even">
<td align="left">ENSG00000114770</td>
<td align="left">ABCC5</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000106077</td>
<td align="left">ABHD11</td>
</tr>
</tbody>
</table>
<p>How many genes in total, are included in the <code>gene_list</code>
object?</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">nrow</span>(gene_list)</span></code></pre></div>
<pre><code>## [1] 1900</code></pre>
<p>The <code>signatures</code> object has all the information on what
genes are associated with what signature. This object is a list of 6
data frames, for different signatures. Typically, the columns of each
data frame are the following; <code>hgnc_symbol</code>,
<code>ensembl_gene_id</code> and <code>signature</code>. Let’s have a
look at the immune signatures object in this list.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">table</span>(signatures<span class="sc">$</span>immune<span class="sc">$</span>signature), </span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 3. Immune Signatures&quot;</span>, </span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>             <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;Signature&quot;</span>, <span class="st">&quot;Genes (n)&quot;</span>))</span></code></pre></div>
<table>
<caption>Table 3. Immune Signatures</caption>
<thead>
<tr class="header">
<th align="left">Signature</th>
<th align="right">Genes (n)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">B-cells</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">Cytotoxicity Score</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">Endothelial cells</td>
<td align="right">12</td>
</tr>
<tr class="even">
<td align="left">Fibroblasts</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">M2 macrophage</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">Macrophages</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">Monocytic lineage</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">Myeloid Dendritic Cells</td>
<td align="right">5</td>
</tr>
<tr class="odd">
<td align="left">Neutrophils</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">NK-cells</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">Smooth muscle</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">T-cells</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">T-cells CD8+</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="subtype-prediction" class="section level2">
<h2>Subtype Prediction</h2>
<p>The main function in this package is
<code>lundtax_predict_sub</code>. This function makes prediction calls
based on gene expression data. It also internally calls
<code>lundtax_calc_sigscore</code> to calculate signature scores for a
variety of,for the subtype, relevant signature scores. The main input
for this function is a data object with expression values, given to the
<code>this_data</code> parameter. This should be a data frame or matrix
with expression values, gene identifiers in as row names and samples in
columns. The function supports gene identifiers in one of the following
formats; <strong>HGNC symbols</strong> (default) or <strong>Ensembl Gene
IDs</strong>. The format for the provided gene identifiers should be
given to the <code>gene_id</code> parameter. This function also depends
on the <code>predict_RF</code> function from
<code>multiclassPairs</code>. Thus, the function can also take arguments
from this function. For example, if required genes are missing from the
input data, the user can run the prediction function with
<code>impute = TRUE</code>.</p>
<div id="example-1---predict-subtype" class="section level3">
<h3>Example 1 - Predict Subtype</h3>
<p>In this first example we will call this function on the bundled
expression data to determine a subtype for each of the included samples.
Note, that we run this example with <code>impute = TRUE</code> to
compensate for missing genes in the demo data set. For transparency, all
other parameters are called with default parameters for demonstrating
purposes.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>sjodahl_predicted <span class="ot">=</span> <span class="fu">lundtax_predict_sub</span>(<span class="at">this_data =</span> sjodahl_2017, </span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>                                        <span class="at">gene_id =</span> <span class="st">&quot;hgnc_symbol&quot;</span>,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                                        <span class="at">impute =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                                        <span class="at">threshold_progression =</span> <span class="fl">0.58</span>,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>                                        <span class="at">log_transform =</span> <span class="cn">TRUE</span>, </span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>                                        <span class="at">adjust =</span> <span class="cn">FALSE</span>, </span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>                                        <span class="at">adj_factor =</span> <span class="fl">5.1431</span>, </span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>                                        <span class="at">impute_reject =</span> <span class="fl">0.67</span>, </span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>                                        <span class="at">impute_kNN =</span> <span class="dv">5</span>, </span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>                                        <span class="at">subtype_only =</span> <span class="cn">FALSE</span>, </span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>                                        <span class="at">include_data =</span> <span class="cn">FALSE</span>, </span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>                                        <span class="at">include_pred_scores =</span> <span class="cn">TRUE</span>, </span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>                                        <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Calculating proliferation scores...</code></pre>
<pre><code>## 6 out of 111 genes are missing from the data...</code></pre>
<pre><code>## [1] &quot;NUSAP1&quot;   &quot;PRC1&quot;     &quot;FAM72D&quot;   &quot;FAM72A&quot;   &quot;RNASEH2A&quot; &quot;POLA2&quot;</code></pre>
<pre><code>## Calculating progression scores...</code></pre>
<pre><code>## 9 out of 100 genes are missing from the data...</code></pre>
<pre><code>## [1] &quot;CCDC150&quot;       &quot;TATDN2&quot;        &quot;FAM72A&quot;        &quot;FAM72B&quot;       
## [5] &quot;ZNF706&quot;        &quot;ESPL1&quot;         &quot;LOXL2&quot;         &quot;PCDHB3&quot;       
## [9] &quot;ZNF670-ZNF695&quot;</code></pre>
<pre><code>## 19 out of 100 genes are missing from the data...</code></pre>
<pre><code>##  [1] &quot;UBE2D3&quot;  &quot;TSPYL1&quot;  &quot;EEF1B2&quot;  &quot;LY75&quot;    &quot;RPL26&quot;   &quot;EEF1A1&quot;  &quot;ZNF816&quot; 
##  [8] &quot;TMBIM4&quot;  &quot;RPL27A&quot;  &quot;PFDN5&quot;   &quot;RPS27&quot;   &quot;RPS3A&quot;   &quot;EIF4B&quot;   &quot;TPT1&quot;   
## [15] &quot;SLC35A1&quot; &quot;RPL13A&quot;  &quot;LDHB&quot;    &quot;EEF1G&quot;   &quot;BORCS7&quot;</code></pre>
<pre><code>## Calculating immune scores...</code></pre>
<pre><code>## 3 out of 71 genes are missing from the data...</code></pre>
<pre><code>## [1] &quot;KIR2DL3&quot; &quot;KLRK1&quot;   &quot;FCGR3B&quot;</code></pre>
<pre><code>## Calculating 141 scores...</code></pre>
<pre><code>## 30 out of 282 genes are missing from the data...</code></pre>
<pre><code>##  [1] &quot;FYB&quot;      &quot;EVI2B&quot;    &quot;IL2RG&quot;    &quot;MAFB&quot;     &quot;HLA-DPA1&quot; &quot;HLA-DPB1&quot;
##  [7] &quot;IFI30&quot;    &quot;PVRIG&quot;    &quot;HLA-E&quot;    &quot;PTPRCAP&quot;  &quot;HLA-B&quot;    &quot;HLA-DMA&quot; 
## [13] &quot;KLRK1&quot;    &quot;HLA-DRA&quot;  &quot;HLA-F&quot;    &quot;HLA-G&quot;    &quot;CD302&quot;    &quot;FCGR2A&quot;  
## [19] &quot;WISP1&quot;    &quot;GPR124&quot;   &quot;LPPR4&quot;    &quot;GIMAP5&quot;   &quot;LMOD1&quot;    &quot;MXRA5&quot;   
## [25] &quot;CH25H&quot;    &quot;COL14A1&quot;  &quot;CD248&quot;    &quot;IL18R1&quot;   &quot;TXNDC3&quot;   &quot;ODZ4&quot;</code></pre>
<pre><code>## Merging scores...</code></pre>
<p>Let’s explore what the returned object looks like. The return is a
lsit of 4, with the following information in it:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">summary</span>(sjodahl_predicted), </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 4. Prediction Calls&quot;</span>, </span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>             <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;Object&quot;</span>, <span class="st">&quot;Length&quot;</span>, <span class="st">&quot;Class&quot;</span>, <span class="st">&quot;Mode&quot;</span>))</span></code></pre></div>
<table>
<caption>Table 4. Prediction Calls</caption>
<thead>
<tr class="header">
<th align="left">Object</th>
<th align="left">Length</th>
<th align="left">Class</th>
<th align="left">Mode</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">subtype_scores</td>
<td align="left">2136</td>
<td align="left">-none-</td>
<td align="left">numeric</td>
</tr>
<tr class="even">
<td align="left">predictions_7classes</td>
<td align="left">267</td>
<td align="left">-none-</td>
<td align="left">character</td>
</tr>
<tr class="odd">
<td align="left">predictions_5classes</td>
<td align="left">267</td>
<td align="left">-none-</td>
<td align="left">character</td>
</tr>
<tr class="even">
<td align="left">scores</td>
<td align="left">35</td>
<td align="left">data.frame</td>
<td align="left">list</td>
</tr>
</tbody>
</table>
<p>The <code>subtype_scores</code> object holds all the subtype
prediction scores for all included samples.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sjodahl_predicted<span class="sc">$</span>subtype_scores, <span class="dv">5</span>), </span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 5. Prediction Scores&quot;</span>,)</span></code></pre></div>
<table>
<caption>Table 5. Prediction Scores</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Uro</th>
<th align="right">UroA</th>
<th align="right">UroB</th>
<th align="right">UroC</th>
<th align="right">GU</th>
<th align="right">BaSq</th>
<th align="right">Mes</th>
<th align="right">ScNE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1.CEL</td>
<td align="right">0.9924</td>
<td align="right">0.9350</td>
<td align="right">0.0048</td>
<td align="right">0.0602</td>
<td align="right">0.0072</td>
<td align="right">0.0004</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
</tr>
<tr class="even">
<td align="left">2.CEL</td>
<td align="right">0.0016</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.0052</td>
<td align="right">0.0054</td>
<td align="right">0.9804</td>
<td align="right">0.0074</td>
</tr>
<tr class="odd">
<td align="left">3.CEL</td>
<td align="right">0.0682</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.7754</td>
<td align="right">0.0834</td>
<td align="right">0.0390</td>
<td align="right">0.0340</td>
</tr>
<tr class="even">
<td align="left">4.CEL</td>
<td align="right">0.9930</td>
<td align="right">0.9334</td>
<td align="right">0.0028</td>
<td align="right">0.0638</td>
<td align="right">0.0066</td>
<td align="right">0.0004</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
</tr>
<tr class="odd">
<td align="left">5.CEL</td>
<td align="right">0.9808</td>
<td align="right">0.0058</td>
<td align="right">0.9878</td>
<td align="right">0.0064</td>
<td align="right">0.0076</td>
<td align="right">0.0090</td>
<td align="right">0.0006</td>
<td align="right">0.0020</td>
</tr>
</tbody>
</table>
<p>The second and third element in the list has the 7 class and 5 class
prediction calls respectivly.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">cbind</span>(<span class="fu">head</span>(sjodahl_predicted<span class="sc">$</span>predictions_7classes, <span class="dv">5</span>),</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>                   <span class="fu">head</span>(sjodahl_predicted<span class="sc">$</span>predictions_5classes, <span class="dv">5</span>)), </span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 6. Prediction Classes Combined&quot;</span>,</span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>             <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;Sample ID&quot;</span>, <span class="st">&quot;7 Class&quot;</span>, <span class="st">&quot;5 Class&quot;</span>))</span></code></pre></div>
<table>
<caption>Table 6. Prediction Classes Combined</caption>
<thead>
<tr class="header">
<th align="left">Sample ID</th>
<th align="left">7 Class</th>
<th align="left">5 Class</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1.CEL</td>
<td align="left">UroA</td>
<td align="left">Uro</td>
</tr>
<tr class="even">
<td align="left">2.CEL</td>
<td align="left">Mes</td>
<td align="left">Mes</td>
</tr>
<tr class="odd">
<td align="left">3.CEL</td>
<td align="left">GU</td>
<td align="left">GU</td>
</tr>
<tr class="even">
<td align="left">4.CEL</td>
<td align="left">UroA</td>
<td align="left">Uro</td>
</tr>
<tr class="odd">
<td align="left">5.CEL</td>
<td align="left">UroB</td>
<td align="left">Uro</td>
</tr>
</tbody>
</table>
<p>The last object in the returned list has all the signature scores,
let’s start with printing the available variables in this object</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">colnames</span>(sjodahl_predicted<span class="sc">$</span>scores)</span></code></pre></div>
<pre><code>##  [1] &quot;proliferation_score&quot;                &quot;progression_score&quot;                 
##  [3] &quot;progression_risk&quot;                   &quot;molecular_grade_who_2016_score&quot;    
##  [5] &quot;molecular_grade_who_2016&quot;           &quot;molecular_grade_who_1999_score&quot;    
##  [7] &quot;molecular_grade_who_1999&quot;           &quot;stromal141_up&quot;                     
##  [9] &quot;immune141_up&quot;                       &quot;b_cells&quot;                           
## [11] &quot;b_cells_proportion&quot;                 &quot;t_cells&quot;                           
## [13] &quot;t_cells_proportion&quot;                 &quot;t_cells_cd8&quot;                       
## [15] &quot;t_cells_cd8_proportion&quot;             &quot;nk_cells&quot;                          
## [17] &quot;nk_cells_proportion&quot;                &quot;cytotoxicity_score&quot;                
## [19] &quot;cytotoxicity_score_proportion&quot;      &quot;neutrophils&quot;                       
## [21] &quot;neutrophils_proportion&quot;             &quot;monocytic_lineage&quot;                 
## [23] &quot;monocytic_lineage_proportion&quot;       &quot;macrophages&quot;                       
## [25] &quot;macrophages_proportion&quot;             &quot;m2_macrophage&quot;                     
## [27] &quot;m2_macrophage_proportion&quot;           &quot;myeloid_dendritic_cells&quot;           
## [29] &quot;myeloid_dendritic_cells_proportion&quot; &quot;endothelial_cells&quot;                 
## [31] &quot;endothelial_cells_proportion&quot;       &quot;fibroblasts&quot;                       
## [33] &quot;fibroblasts_proportion&quot;             &quot;smooth_muscle&quot;                     
## [35] &quot;smooth_muscle_proportion&quot;</code></pre>
<p>Let’s also print out the first 5 rows and 3 columns of this
object.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(sjodahl_predicted<span class="sc">$</span>scores[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], </span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>             <span class="at">caption =</span> <span class="st">&quot;Table 7. Signature Score Subset&quot;</span>)</span></code></pre></div>
<table>
<caption>Table 7. Signature Score Subset</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">proliferation_score</th>
<th align="right">progression_score</th>
<th align="left">progression_risk</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1.CEL</td>
<td align="right">0.7310869</td>
<td align="right">0.7675475</td>
<td align="left">HR</td>
</tr>
<tr class="even">
<td align="left">2.CEL</td>
<td align="right">1.0419366</td>
<td align="right">0.9923554</td>
<td align="left">HR</td>
</tr>
<tr class="odd">
<td align="left">3.CEL</td>
<td align="right">0.4846604</td>
<td align="right">0.5824121</td>
<td align="left">HR</td>
</tr>
<tr class="even">
<td align="left">4.CEL</td>
<td align="right">0.6596973</td>
<td align="right">0.7804901</td>
<td align="left">HR</td>
</tr>
<tr class="odd">
<td align="left">5.CEL</td>
<td align="right">0.8079496</td>
<td align="right">0.6303658</td>
<td align="left">HR</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>This part of the vignette will demonstrate how to integrate the
plotting functions to the return of the main functions in this package.
Currently, there are three alternatives for visualize the results;
<code>plot_hm_signatures</code>, <code>plot_hm_score</code>,
<code>plot_ranked_score</code>. The first of these is developed to plots
a heatmap, including genes and signatures of interest, with prediction
results and scores on top. Let’s call this function on the generated
prediction call set in the previous example.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">plot_hm_signatures</span>(<span class="at">these_predictions =</span> sjodahl_predicted, </span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>                   <span class="at">this_data =</span> sjodahl_2017,</span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>                   <span class="at">subtype_annotation =</span> <span class="st">&quot;7_class&quot;</span>)</span></code></pre></div>
<pre><code>## No out_path provided, the function will return the heatmap within your R session...</code></pre>
<pre><code>## null device 
##           1</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
